File: packages/cli/src/commands/pack.ts
import { Command } from 'commander';
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { ManifestSchema } from '@spm/shared';
import type { Manifest } from '@spm/shared';
import * as tar from 'tar';
import { icons, c, log, logJson, getOutputMode, logError, withSpinner } from '../lib/output.js';

// -- Helpers --

const formatBytes = (bytes: number): string => {
  if (bytes < 1024) return `${bytes} B`;
  const kb = bytes / 1024;
  if (kb < 1024) return `${kb.toFixed(1)} KB`;
  const mb = kb / 1024;
  return `${mb.toFixed(1)} MB`;
};

const collectScriptFiles = async (dir: string, cwd: string): Promise<string[]> => {
  const files: string[] = [];
  const entries = await fs.readdir(dir);
  for (const name of entries) {
    const fullPath = path.join(dir, name);
    const stat = await fs.stat(fullPath);
    if (stat.isDirectory()) {
      const sub = await collectScriptFiles(fullPath, cwd);
      files.push(...sub);
    } else {
      files.push(path.relative(cwd, fullPath));
    }
  }
  return files;
};

const collectPackFiles = async (cwd: string, manifest: Manifest): Promise<string[]> => {
  const files: string[] = [];

  // Always include manifest.json
  files.push('manifest.json');

  // Always include SKILL.md if it exists
  try {
    await fs.access(path.join(cwd, 'SKILL.md'));
    files.push('SKILL.md');
  } catch {
    // SKILL.md is optional but recommended
  }

  // Include scripts/ directory
  const scriptsDir = path.join(cwd, 'scripts');
  try {
    const scriptFiles = await collectScriptFiles(scriptsDir, cwd);
    files.push(...scriptFiles);
  } catch {
    // scripts/ not required
  }

  // Include files from manifest.files.include
  if (manifest.files?.include) {
    for (const pattern of manifest.files.include) {
      // Simple glob: just add the file if it exists
      try {
        await fs.access(path.join(cwd, pattern));
        if (!files.includes(pattern)) {
          files.push(pattern);
        }
      } catch {
        // File doesn't exist, skip
      }
    }
  }

  return files;
};

// -- Command --

export const registerPackCommand = (program: Command): void => {
  program
    .command('pack')
    .description('Create a .skl archive from the current skill')
    .action(async () => {
      const mode = getOutputMode();
      if (mode === 'silent') return;

      try {
        const cwd = process.cwd();
        const manifestPath = path.join(cwd, 'manifest.json');

        // Read and validate manifest
        let rawManifest: string;
        try {
          rawManifest = await fs.readFile(manifestPath, 'utf-8');
        } catch {
          logError(
            'No manifest.json found',
            'This command must be run in a skill directory.',
            `Run ${c.cmd('spm init')} to create a new skill.`,
          );
          process.exitCode = 1;
          return;
        }

        let parsedJson: unknown;
        try {
          parsedJson = JSON.parse(rawManifest);
        } catch {
          logError('Invalid manifest.json', 'File is not valid JSON.');
          process.exitCode = 1;
          return;
        }

        const result = ManifestSchema.safeParse(parsedJson);
        if (!result.success) {
          const issues = result.error.issues
            .map((i) => `  ${i.path.join('.')}: ${i.message}`)
            .join('\n');
          logError('Invalid manifest.json', `Validation errors:\n${issues}`);
          process.exitCode = 1;
          return;
        }

        const manifest = result.data;
        const archiveName = `${manifest.name.replace('/', '-')}-${manifest.version}.skl`;
        const archivePath = path.join(cwd, archiveName);

        // Collect files
        const files = await collectPackFiles(cwd, manifest);

        if (files.length === 0) {
          logError('No files to pack', 'The skill directory appears to be empty.');
          process.exitCode = 1;
          return;
        }

        // Create tar.gz archive
        await withSpinner(
          `Packing ${c.name(`${manifest.name}@${manifest.version}`)}...`,
          async () => {
            await tar.create(
              {
                gzip: true,
                file: archivePath,
                cwd,
              },
              files,
            );
          },
        );

        // Get archive size
        const stat = await fs.stat(archivePath);

        if (mode === 'json') {
          logJson({
            name: manifest.name,
            version: manifest.version,
            archive: archiveName,
            files: files.length,
            size: stat.size,
          });
          return;
        }

        log('');
        log(
          `${icons.success} Packed ${c.name(`${manifest.name}@${manifest.version}`)} (${files.length} file${files.length === 1 ? '' : 's'}, ${formatBytes(stat.size)})`,
        );
        log('');
        for (const file of files) {
          log(`  ${c.dim(file)}`);
        }
        log('');
        log(`  ${c.dim('Archive:')} ${c.path(archiveName)}`);
        log(`  ${c.dim('Publish:')} ${c.cmd('spm publish')}`);
        log('');
      } catch (err) {
        logError('Pack failed', err instanceof Error ? err.message : String(err));
        process.exitCode = 1;
      }
    });
};

File: cli-go/cmd/pack.go
package cmd

import (
	"archive/tar"
	"compress/gzip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/skillpkg/spm/internal/manifest"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var packCmd = &cobra.Command{
	Use:   "pack [path]",
	Short: "Create a .skl archive from the current skill",
	Long:  "Pack a skill directory into a .skl file (tar.gz archive) containing manifest.json, SKILL.md, and declared files.",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runPack,
}

func init() {
	rootCmd.AddCommand(packCmd)
}

func runPack(_ *cobra.Command, args []string) error {
	dir, err := resolveSkillDir(args)
	if err != nil {
		return err
	}

	// Read and validate manifest
	manifestPath := filepath.Join(dir, "manifest.json")
	m, err := manifest.LoadFile(manifestPath)
	if err != nil {
		return fmt.Errorf("could not load manifest.json: %w", err)
	}
	if err := manifest.Validate(m); err != nil {
		return fmt.Errorf("manifest validation failed: %w", err)
	}

	// Collect files to pack
	files, err := collectPackFiles(dir, m)
	if err != nil {
		return fmt.Errorf("collecting files: %w", err)
	}
	if len(files) == 0 {
		return fmt.Errorf("no files to pack")
	}

	// Build archive name
	archiveName := fmt.Sprintf("%s-%s.skl", strings.ReplaceAll(m.Name, "/", "-"), m.Version)
	archivePath := filepath.Join(dir, archiveName)

	// Create tar.gz
	sp := Out.StartSpinner(fmt.Sprintf("Packing %s@%s...", m.Name, m.Version))
	err = createTarGz(archivePath, dir, files)
	output.StopSpinner(sp)
	if err != nil {
		return fmt.Errorf("creating archive: %w", err)
	}

	// Get size
	stat, err := os.Stat(archivePath)
	if err != nil {
		return fmt.Errorf("reading archive size: %w", err)
	}

	// JSON mode
	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(map[string]any{
			"name":    m.Name,
			"version": m.Version,
			"archive": archiveName,
			"path":    archivePath,
			"files":   files,
			"size":    stat.Size(),
		})
	}

	// Human mode
	Out.Log("")
	Out.Log("  %s Packed %s (%d file%s, %s)",
		output.Icons["success"],
		output.Cyan(fmt.Sprintf("%s@%s", m.Name, m.Version)),
		len(files), pluralS(len(files)),
		formatBytes(stat.Size()),
	)
	Out.Log("")
	for _, f := range files {
		Out.Log("  %s", output.Dim(f))
	}
	Out.Log("")
	Out.Log("  %s %s", output.Dim("Archive:"), output.Cyan(archiveName))
	Out.Log("  %s %s", output.Dim("Publish:"), output.Cyan("spm publish"))
	Out.Log("")

	return nil
}

func collectPackFiles(dir string, m *manifest.Manifest) ([]string, error) {
	var files []string

	// Always include manifest.json
	files = append(files, "manifest.json")

	// Include SKILL.md if it exists
	if _, err := os.Stat(filepath.Join(dir, "SKILL.md")); err == nil {
		files = append(files, "SKILL.md")
	}

	// Include scripts/ directory
	scriptsDir := filepath.Join(dir, "scripts")
	if info, err := os.Stat(scriptsDir); err == nil && info.IsDir() {
		err := filepath.Walk(scriptsDir, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() {
				rel, err := filepath.Rel(dir, path)
				if err != nil {
					return err
				}
				files = append(files, rel)
			}
			return nil
		})
		if err != nil {
			return nil, fmt.Errorf("walking scripts: %w", err)
		}
	}

	// Include files from manifest.files.include
	if m.Files != nil {
		for _, pattern := range m.Files.Include {
			fullPath := filepath.Join(dir, pattern)
			if _, err := os.Stat(fullPath); err == nil {
				found := false
				for _, f := range files {
					if f == pattern {
						found = true
						break
					}
				}
				if !found {
					files = append(files, pattern)
				}
			}
		}
	}

	return files, nil
}

func createTarGz(archivePath, baseDir string, files []string) error {
	outFile, err := os.Create(archivePath)
	if err != nil {
		return err
	}
	defer outFile.Close()

	gzWriter := gzip.NewWriter(outFile)
	defer gzWriter.Close()

	tw := tar.NewWriter(gzWriter)
	defer tw.Close()

	for _, file := range files {
		fullPath := filepath.Join(baseDir, file)
		info, err := os.Stat(fullPath)
		if err != nil {
			return fmt.Errorf("stat %s: %w", file, err)
		}

		header, err := tar.FileInfoHeader(info, "")
		if err != nil {
			return fmt.Errorf("header %s: %w", file, err)
		}
		header.Name = file

		if err := tw.WriteHeader(header); err != nil {
			return fmt.Errorf("write header %s: %w", file, err)
		}

		f, err := os.Open(fullPath)
		if err != nil {
			return fmt.Errorf("open %s: %w", file, err)
		}
		if _, err := io.Copy(tw, f); err != nil {
			f.Close()
			return fmt.Errorf("copy %s: %w", file, err)
		}
		f.Close()
	}

	return nil
}

func formatBytes(b int64) string {
	if b < 1024 {
		return fmt.Sprintf("%d B", b)
	}
	kb := float64(b) / 1024
	if kb < 1024 {
		return fmt.Sprintf("%.1f KB", kb)
	}
	mb := kb / 1024
	return fmt.Sprintf("%.1f MB", mb)
}
